src/typechecker.erl

```diff
@@ -1968,44 +2013,151 @@ do_type_check_expr_in(Env, ResTy, {'try', _, Block, CaseCs, CatchCs, AfterBlock}
-	false ->
-	  throw({type_error, arith_error, Op, P, ResTy})
-    end.
+    type_check_arith_op_in(Env, number, ResTy, Op, P, Arg1, Arg2).
+
 type_check_int_op_in(Env, ResTy, Op, P, Arg1, Arg2) ->
-    case subtype(ResTy, {type, erl_anno:new(0), 'integer', []}, Env#env.tenv) of
-	{true, Cs} ->
-	  {VarBinds1, Cs1} = type_check_expr_in(Env, ResTy, Arg1),
-	  {VarBinds2, Cs2} = type_check_expr_in(Env, ResTy, Arg2),
-	  {union_var_binds([VarBinds1, VarBinds2])
-	  ,constraints:combine([Cs, Cs1, Cs2])};
-	false ->
-	  throw({type_error, int_error, Op, P, ResTy})
+    type_check_arith_op_in(Env, integer, ResTy, Op, P, Arg1, Arg2).
+
+type_check_arith_op_in(Env, Kind, ResTy, Op, P, Arg1, Arg2) ->
+    ResTy1 = glb(type(Kind), ResTy, Env#env.tenv),
+    IsNone =
+        case ResTy1 of
+            %% TODO: allow none() if checking against none()? Not clear that
+            %% this is sensible, since in that case you'd like to only require
+            %% *one* of the arguments to be none(), not both.
+            {type, _, none, []} -> true;
+            _ -> false
+        end,
+    case IsNone of
+        false ->
+            case arith_op_arg_types(Op, ResTy1) of
+                {ArgTy1, ArgTy2} ->
+                    {VarBinds1, Cs1} = type_check_expr_in(Env, ArgTy1, Arg1),
+                    {VarBinds2, Cs2} = type_check_expr_in(Env, ArgTy2, Arg2),
+                    {union_var_binds([VarBinds1, VarBinds2]),
+                     constraints:combine([Cs1, Cs2])};
+                false ->
+                    throw({type_error, arith_type_too_precise, Op, P, ResTy1})
+            end;
+        true ->
+            Tag = if Kind == integer -> int_error;
+                     true            -> arith_error end,
+            throw({type_error, Tag, Op, P, ResTy})
     end.
+
+%% What types should be pushed into the arguments if checking an operator
+%% application against a given type.
+
+%% any() is always fine
+arith_op_arg_types(_Op, Ty = {type, _, any, []}) ->
+    {Ty, Ty};
+
+%% '/' can't produce an integer type
+arith_op_arg_types('/', Ty) when ?is_int_type(Ty) ->
+    false;
+
+%% integer() is closed under all operators except '/'
+arith_op_arg_types(_, Ty = {type, _, integer, []}) ->
+    {Ty, Ty};
+
+%% float() is closed under non-integer-exclusive operations and accepts
+%% any number() for '/'.
+%% Some precision lost: if at least one argument is a float() the other can
+%% be integer().
+arith_op_arg_types(Op, Ty = {type, _, float, []}) ->
+    case Op of
+        _ when Op == '+'; Op == '*'; Op == '-' ->
+            {Ty, Ty};
+        '/' -> {type(number), type(number)};
+        _   -> false
+    end;
+
+%% Singleton types are not closed under any operations
+arith_op_arg_types(_, {T, _, _}) when T == integer; T == float ->
+    false;
+
+%% pos_integer() is closed under '+',  '*', and 'bor'
+arith_op_arg_types(Op, Ty = {type, _, pos_integer, []}) ->
+    case lists:member(Op, ['+', '*', 'bor']) of
+        true -> {Ty, Ty};
```

Comment:
What about `neg_integer() * neg_integer() :: pos_integer()`?

Reply:
This would require the same machinery as `integer() * float() :: float()`.